rust/otel-arrow-rust/src/otap/groups.rs

```diff
@@ -1431,144 +1430,386 @@ impl UnifiedDictionaryTypeSelector {
+/// * both type `T` and the underlying element type must impl `ArrowNativeType` since that ensures
+///   that our reinterpretation is safe
+fn visit_primitive_array<T, F>(
+    array: &ArrayRef,
+    mut visitor: F,
+) -> ControlFlow<UnifiedDictionaryType>
+where
+    T: ArrowNativeType,
+    F: FnMut(&[T]) -> ControlFlow<UnifiedDictionaryType>,
+{
+    assert!(array.data_type().is_primitive());
+    assert_eq!(
+        size_of::<T>(),
+        bitmap_data_type_width(array.data_type()).unwrap_or(usize::MAX)
+    );
+
+    // Why the `to_data` conversion? I think this is the only safe way to allow us to
+    // interpret the underlying byte array as an array of another type with the same size
+    // and compatible alignment. Doing that means we only need to cover three cases (1-byte,
+    // 2-bytes, 4-bytes) instead of many more cases.
+    let data = array.to_data();
+    // We're relying on the fact that `PrimitiveArray`s have one and only one buffer, so it
+    // is safe to just look at buffer 0. They kind of have to since `values()` returns a
+    // single `ScalarBuffer` that behaves likes a `[T]` so Arrow can't change the
+    // `PrimitiveArray` has only one buffer property without an API break.
+    assert_eq!(data.buffers().len(), 1);
+    let values = data.buffer::<T>(0);
+    match (array.null_count(), array.nulls()) {
+        (0, _) => {
+            visitor(values)?;
+        }
+        (_, Some(nulls)) => {
+            for (start, end) in nulls.valid_slices() {
+                let range = start..end;
+                visitor(&values[range])?;
+            }
+        }
+        (1.., None) => unreachable!(),
+    }
+    ControlFlow::Continue(())
 }
 
-fn access_fsb_dict_values(fsb_array: &FixedSizeBinaryArray, index: usize) -> Option<&[u8]> {
-    fsb_array.is_valid(index).then(|| fsb_array.value(index))
+struct LargeCardinalityEstimator<'parent> {
+    // these two are immutable inputs
+    arrays: &'parent [ArrayRef],
+    smallest_key_type: &'parent DataType,
+
+    // these are immutable after construction time
+    total_elements: usize, // Note: we only consider non-null items!
+
+    // and these are mutable state variables
+    elements_seen: usize,
+    dedup: AHashSet<&'parent [u8]>,
 }
 
-fn access_primitive_dict_values<T: ArrowPrimitiveType>(
-    array: &PrimitiveArray<T>,
-    index: usize,
-) -> Option<&[u8]> {
-    array.is_valid(index).then(|| {
-        let size = size_of::<T::Native>();
-        let values_buffer = array.values().inner();
-        let offset = index * size;
-        &values_buffer[offset..offset + size]
-    })
+impl<'parent> LargeCardinalityEstimator<'parent> {
+    fn new(arrays: &'parent [ArrayRef], smallest_key_type: &'parent DataType) -> Self {
+        let total_elements: usize = arrays.iter().map(|v| v.len() - v.null_count()).sum();
+        let elements_seen = 0;
+
+        let dedup = AHashSet::with_capacity(if total_elements <= u8::MAX as usize {
+            u8::MAX as usize
```

Comment:
Why not just use `total_elements` as the capacity?

```suggestion
           total_elements
```